### DISCUSSION TITLE
Error message rheoFoam multi-mode turbulence

### DISCUSSION THREAD

**Discussion by nk583** [fppimenta/rheoTool#59] [Q&A]
Hi,

I have been using the rheoTool toolbox to perform some DNS simulations of turbulent flow over a backwards-facing step, but my simulation is crashing, and I am ending up with a long error message that I am not sure what it is telling me.
[rheoFoam error.txt](https://github.com/fppimenta/rheoTool/files/6617421/rheoFoam.error.txt)

I am using a four-mode PTT-Linear Log model for my simulation (for blood flow). I can see from the log file that the iteration before it fails, the time step continuity errors get high, and it crashes.  

Any help would be great.

**Comment by fppimenta** [comment]
What you are getting is simply a blowup, i.e. your current settings render the simulation numerically unstable. Some comments:
  - if your case is transient, the final residual for pressure is too high for accurate results (rtol and atol). If the case is steady, accuracy is not an issue, but they can still be too high to ensure stability.
  -  if your case is transient, your max Courant number is likely too high to ensure accuracy. If it's steady, it can still be too high to keep stability. Compare it with all the time scales of your case (viscous, elastic, inertial).
  -  if Wi is high, the case is naturally stiff and prone to problems.
  - start lowering the time-step and tightening the residuals. If not enough, play with discretization schemes, BC, rheological model, etc. I am assuming the mesh is ok, if not then it might be the source of the blowup.

**Reply by nk583** [reply]
Thanks for the suggestions. What I interpreted from the error message is that it might have something to do with theta, but if you think it might be an issue with general numerical stability I can look into that a bit more. 

The case is transient, my pressure reltol 0.05, I will reduce that tolerance lower. My Wi is not extremely high, however, I am using the Log model to help with this. 

I have been working on this for a while for different Re and they seemed to all blow-up after about 1.5 flow-through times, and I have always thought it something to do with the constitutive model parameters. I do normally have the auto-adjust timestep on, when this used the time step gets incredibly small (1e-9), which is not feasible. The log file I sent was for testing something, so I set the Co No. a little higher, but I do normally have it < 0.25. 

I have the same settings for Newtonian and shear-thinning rheology (albeit not using rheoFoam), so I am fairly sure the BCs, discretization and mesh are satisfactory. 

Do you have any recommendation for BCs and discretization schemes for tau? I have been using Gauss DefCmpw Cubista.

**Comment by fppimenta** [comment]
. If your case is transient use abstol for pressure instead of rtol. Say 1e-7 or lower.
. Don't use Co as reference if you have more stringent time-scales. Is your usual time-step at least lower than lambda/1000 (and tdiff/1000 if diffusion dominated)?
. If the Wi is not very high try a non-Log version, because they are more implicit than the log ones. What happens at lower Wi and/or with less modes (one of the modes can be particularly problematic) or other constitutive models?
. There are several things that can be tried with schemes, BCs, meshes, stabilization, etc. but they are case-dependent. There is no magic bullet.
. Don't change everything at the same time.

**Reply by nk583** [reply]
For the viscoelastic I am using, Mode 1 has the longest lambda and is 7s, whilst Mode 4 has the shortest is 0.006s. My timestep is normally smaller than 1e-5, maybe an even smaller timestep is needed. For my Re that timestep is satisfactory, but you are correct that I need to take into account the elastic time scales also. 

I have performed some simulations of turbulent channel flow with the non-Log version, and I have found that using Modes 2, 3 and 4 tend to solve fine and I get decent results. When I just use Mode 1 on its own, or with the other 3 modes the flow field relaminarizes. 

I think the 1st mode is the one that is causing the issue. I have not tried it with other constitutive models, I can try with the Giesekus model as I have the parameters for that.
